File: src/memory_service/store.py
"""SQLite persistence layer.

Why SQLite (baseline): zero external services, single file on a Docker volume =>
real persistence with no infra; synchronous reads-after-writes by construction
(no eventual consistency); trivially fast to iterate. The store is deliberately
behind a small method surface so we can swap in Postgres+pgvector / Qdrant as a
benchmarked variant later without touching the API or pipelines.

Concurrency: WAL mode for concurrent readers; a process-level lock serializes
writes (SQLite allows a single writer). Each operation uses its own short-lived
connection so FastAPI's threadpool workers don't share cursors.
"""

from __future__ import annotations

import json
import sqlite3
import threading
import uuid
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from .extraction.base import ExtractedMemory
from .text import token_set

_SCHEMA = """
CREATE TABLE IF NOT EXISTS turns (
    id          TEXT PRIMARY KEY,
    session_id  TEXT NOT NULL,
    user_id     TEXT,
    timestamp   TEXT,
    messages    TEXT NOT NULL,
    metadata    TEXT,
    text        TEXT NOT NULL DEFAULT '',
    created_at  TEXT NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_turns_session ON turns(session_id);
CREATE INDEX IF NOT EXISTS idx_turns_user    ON turns(user_id);

CREATE TABLE IF NOT EXISTS memories (
    id             TEXT PRIMARY KEY,
    user_id        TEXT,
    session_id     TEXT,
    type           TEXT NOT NULL,
    key            TEXT NOT NULL,
    value          TEXT NOT NULL,
    confidence     REAL NOT NULL DEFAULT 0.7,
    source_session TEXT,
    source_turn    TEXT,
    created_at     TEXT NOT NULL,
    updated_at     TEXT NOT NULL,
    supersedes     TEXT,
    active         INTEGER NOT NULL DEFAULT 1
);
CREATE INDEX IF NOT EXISTS idx_mem_user ON memories(user_id);
CREATE INDEX IF NOT EXISTS idx_mem_key  ON memories(user_id, key, active);
"""


def _now_iso() -> str:
    return datetime.now(UTC).isoformat()


def _new_id(prefix: str) -> str:
    return f"{prefix}_{uuid.uuid4().hex[:24]}"


def _norm(value: str) -> str:
    return " ".join(value.lower().split())


@dataclass
class TurnRecord:
    id: str
    session_id: str
    user_id: str | None
    timestamp: str | None
    text: str
    metadata: dict[str, Any]


class Store:
    def __init__(self, db_path: str) -> None:
        self.db_path = db_path
        self._write_lock = threading.Lock()
        if db_path != ":memory:":
            Path(db_path).expanduser().parent.mkdir(parents=True, exist_ok=True)

    # -- connection ------------------------------------------------------- #
    def _connect(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self.db_path, timeout=30.0, check_same_thread=False)
        conn.row_factory = sqlite3.Row
        conn.execute("PRAGMA journal_mode=WAL")
        conn.execute("PRAGMA busy_timeout=30000")
        return conn

    def init(self) -> None:
        with self._write_lock, self._connect() as conn:
            conn.executescript(_SCHEMA)

    # -- writes ----------------------------------------------------------- #
    def insert_turn(
        self,
        *,
        session_id: str,
        user_id: str | None,
        messages: list[dict[str, Any]],
        timestamp: str | None,
        metadata: dict[str, Any],
    ) -> str:
        turn_id = _new_id("turn")
        text = "\n".join(
            f"{m.get('role', '')}: {m.get('content', '')}".strip() for m in messages
        ).strip()
        with self._write_lock, self._connect() as conn:
            conn.execute(
                "INSERT INTO turns (id, session_id, user_id, timestamp, messages, metadata, text, created_at)"
                " VALUES (?,?,?,?,?,?,?,?)",
                (
                    turn_id,
                    session_id,
                    user_id,
                    timestamp,
                    json.dumps(messages, ensure_ascii=False),
                    json.dumps(metadata, ensure_ascii=False),
                    text,
                    _now_iso(),
                ),
            )
        return turn_id

    def add_memory(
        self,
        em: ExtractedMemory,
        *,
        user_id: str | None,
        session_id: str,
        source_turn: str,
    ) -> str:
        """Persist an extracted memory with fact-evolution semantics.

        - mutable (single-valued) slot: if an active memory exists for the same
          (user_id, key) with a *different* value, supersede it (active=0) and
          insert the new one with ``supersedes`` pointing at the old id. Same
          value => just bump confidence/updated_at (dedupe).
        - additive slot: dedupe on (key, value); otherwise insert.
        """
        now = _now_iso()
        with self._write_lock, self._connect() as conn:
            cur = conn.execute(
                "SELECT * FROM memories WHERE user_id IS ? AND key=? AND active=1 ORDER BY updated_at DESC",
                (user_id, em.key),
            )
            existing = cur.fetchall()

            if em.mutable:
                match = existing[0] if existing else None
                if match is not None:
                    if _norm(match["value"]) == _norm(em.value):
                        conn.execute(
                            "UPDATE memories SET updated_at=?, confidence=MAX(confidence, ?) WHERE id=?",
                            (now, em.confidence, match["id"]),
                        )
                        return str(match["id"])
                    conn.execute(
                        "UPDATE memories SET active=0, updated_at=? WHERE id=?",
                        (now, match["id"]),
                    )
                    return self._insert_memory(
                        conn,
                        em,
                        user_id,
                        session_id,
                        source_turn,
                        now,
                        supersedes=str(match["id"]),
                    )
            else:
                for row in existing:
                    if _norm(row["value"]) == _norm(em.value):
                        conn.execute(
                            "UPDATE memories SET updated_at=?, confidence=MAX(confidence, ?) WHERE id=?",
                            (now, em.confidence, row["id"]),
                        )
                        return str(row["id"])

            return self._insert_memory(
                conn, em, user_id, session_id, source_turn, now, supersedes=None
            )

    @staticmethod
    def _insert_memory(
        conn: sqlite3.Connection,
        em: ExtractedMemory,
        user_id: str | None,
        session_id: str,
        source_turn: str,
        now: str,
        *,
        supersedes: str | None,
    ) -> str:
        mem_id = _new_id("mem")
        conn.execute(
            "INSERT INTO memories (id, user_id, session_id, type, key, value, confidence,"
            " source_session, source_turn, created_at, updated_at, supersedes, active)"
            " VALUES (?,?,?,?,?,?,?,?,?,?,?,?,1)",
            (
                mem_id,
                user_id,
                session_id,
                em.type,
                em.key,
                em.value,
                em.confidence,
                session_id,
                source_turn,
                now,
                now,
                supersedes,
            ),
        )
        return mem_id

    # -- reads ------------------------------------------------------------ #
    def list_memories(self, user_id: str, *, active_only: bool = False) -> list[dict[str, Any]]:
        q = "SELECT * FROM memories WHERE user_id IS ?"
        if active_only:
            q += " AND active=1"
        q += " ORDER BY created_at ASC"
        with self._connect() as conn:
            return [dict(r) for r in conn.execute(q, (user_id,)).fetchall()]

    def superseded_values(self, user_id: str | None, key: str) -> list[str]:
        with self._connect() as conn:
            rows = conn.execute(
                "SELECT value FROM memories WHERE user_id IS ? AND key=? AND active=0"
                " ORDER BY updated_at DESC",
                (user_id, key),
            ).fetchall()
        return [str(r["value"]) for r in rows]

    def recent_turns(
        self, *, session_id: str | None, user_id: str | None, limit: int = 50
    ) -> list[TurnRecord]:
        clauses, params = [], []
        if session_id is not None:
            clauses.append("session_id = ?")
            params.append(session_id)
        elif user_id is not None:
            clauses.append("user_id IS ?")
            params.append(user_id)
        else:
            return []
        where = " WHERE " + " AND ".join(clauses)
        params.append(limit)
        with self._connect() as conn:
            rows = conn.execute(
                f"SELECT * FROM turns{where} ORDER BY created_at DESC LIMIT ?", params
            ).fetchall()
        return [
            TurnRecord(
                id=r["id"],
                session_id=r["session_id"],
                user_id=r["user_id"],
                timestamp=r["timestamp"],
                text=r["text"],
                metadata=json.loads(r["metadata"] or "{}"),
            )
            for r in rows
        ]

    def search(
        self,
        query: str,
        *,
        user_id: str | None,
        session_id: str | None,
        limit: int = 10,
    ) -> list[dict[str, Any]]:
        """Keyword-overlap search across memories + turns (baseline)."""
        qset = token_set(query)
        results: list[dict[str, Any]] = []

        with self._connect() as conn:
            mem_clauses, mem_params = ["active=1"], []
            if user_id is not None:
                mem_clauses.append("user_id IS ?")
                mem_params.append(user_id)
            mrows = conn.execute(
                f"SELECT * FROM memories WHERE {' AND '.join(mem_clauses)}", mem_params
            ).fetchall()

            turn_clauses, turn_params = [], []
            if session_id is not None:
                turn_clauses.append("session_id = ?")
                turn_params.append(session_id)
            elif user_id is not None:
                turn_clauses.append("user_id IS ?")
                turn_params.append(user_id)
            twhere = (" WHERE " + " AND ".join(turn_clauses)) if turn_clauses else ""
            trows = conn.execute(f"SELECT * FROM turns{twhere}", turn_params).fetchall()

        for r in mrows:
            content = f"{r['key']}: {r['value']}"
            score = _overlap(qset, content)
            if score > 0:
                results.append(
                    {
                        "content": content,
                        "score": score,
                        "session_id": r["session_id"],
                        "timestamp": r["updated_at"],
                        "metadata": {"type": r["type"], "memory_id": r["id"]},
                    }
                )
        for r in trows:
            score = _overlap(qset, r["text"])
            if score > 0:
                results.append(
                    {
                        "content": r["text"],
                        "score": score,
                        "session_id": r["session_id"],
                        "timestamp": r["timestamp"],
                        "metadata": json.loads(r["metadata"] or "{}"),
                    }
                )

        results.sort(key=lambda x: x["score"], reverse=True)
        return results[: max(0, limit)]

    # -- deletes ---------------------------------------------------------- #
    def delete_session(self, session_id: str) -> None:
        with self._write_lock, self._connect() as conn:
            conn.execute("DELETE FROM turns WHERE session_id=?", (session_id,))
            conn.execute("DELETE FROM memories WHERE session_id=?", (session_id,))

    def delete_user(self, user_id: str) -> None:
        with self._write_lock, self._connect() as conn:
            conn.execute("DELETE FROM turns WHERE user_id IS ?", (user_id,))
            conn.execute("DELETE FROM memories WHERE user_id IS ?", (user_id,))


def _overlap(qset: set[str], text: str) -> float:
    if not qset:
        return 0.0
    tset = token_set(text)
    if not tset:
        return 0.0
    return round(len(qset & tset) / len(qset), 4)

File: implementations/baseline/src/store.ts
/**
 * SQLite persistence layer (better-sqlite3) — the v0 control store.
 *
 * Why SQLite for the control: zero external services (single file on a Docker
 * volume = real persistence with no infra), and better-sqlite3 is synchronous,
 * so reads-after-writes are correct by construction — no eventual-consistency
 * gap. The store sits behind a small method surface so the exploration branches
 * can swap in pglite/pgvector without touching the API or pipelines.
 *
 * Node runs JS single-threaded, so no write lock is needed; WAL mode keeps reads
 * concurrent with the occasional writer.
 */

import { randomUUID } from "node:crypto";
import { mkdirSync } from "node:fs";
import { dirname } from "node:path";
import Database from "better-sqlite3";
import type { ExtractedMemory } from "./extraction/types";
import { tokenSet } from "./text";

const SCHEMA = `
CREATE TABLE IF NOT EXISTS turns (
  id          TEXT PRIMARY KEY,
  session_id  TEXT NOT NULL,
  user_id     TEXT,
  timestamp   TEXT,
  messages    TEXT NOT NULL,
  metadata    TEXT,
  text        TEXT NOT NULL DEFAULT '',
  created_at  TEXT NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_turns_session ON turns(session_id);
CREATE INDEX IF NOT EXISTS idx_turns_user    ON turns(user_id);

CREATE TABLE IF NOT EXISTS memories (
  id             TEXT PRIMARY KEY,
  user_id        TEXT,
  session_id     TEXT,
  type           TEXT NOT NULL,
  key            TEXT NOT NULL,
  value          TEXT NOT NULL,
  confidence     REAL NOT NULL DEFAULT 0.7,
  source_session TEXT,
  source_turn    TEXT,
  created_at     TEXT NOT NULL,
  updated_at     TEXT NOT NULL,
  supersedes     TEXT,
  active         INTEGER NOT NULL DEFAULT 1
);
CREATE INDEX IF NOT EXISTS idx_mem_user ON memories(user_id);
CREATE INDEX IF NOT EXISTS idx_mem_key  ON memories(user_id, key, active);
`;

export interface MemoryRow {
  id: string;
  user_id: string | null;
  session_id: string | null;
  type: string;
  key: string;
  value: string;
  confidence: number;
  source_session: string | null;
  source_turn: string | null;
  created_at: string;
  updated_at: string;
  supersedes: string | null;
  active: number;
}

export interface TurnRecord {
  id: string;
  session_id: string;
  user_id: string | null;
  timestamp: string | null;
  text: string;
  metadata: Record<string, unknown>;
}

export interface SearchHit {
  content: string;
  score: number;
  session_id: string | null;
  timestamp: string | null;
  metadata: Record<string, unknown>;
}

export interface InsertTurnInput {
  sessionId: string;
  userId: string | null;
  messages: Array<Record<string, unknown>>;
  timestamp: string | null;
  metadata: Record<string, unknown>;
}

const nowIso = (): string => new Date().toISOString();
const newId = (prefix: string): string =>
  `${prefix}_${randomUUID().replace(/-/g, "").slice(0, 24)}`;
const norm = (v: string): string => v.toLowerCase().split(/\s+/).filter(Boolean).join(" ");

function overlapScore(qset: Set<string>, text: string): number {
  if (qset.size === 0) return 0;
  const tset = tokenSet(text);
  if (tset.size === 0) return 0;
  let n = 0;
  for (const t of qset) if (tset.has(t)) n++;
  return Math.round((n / qset.size) * 10000) / 10000;
}

export class Store {
  private db: Database.Database;

  constructor(dbPath: string) {
    if (dbPath !== ":memory:") mkdirSync(dirname(dbPath), { recursive: true });
    this.db = new Database(dbPath);
    this.db.pragma("journal_mode = WAL");
    this.db.pragma("busy_timeout = 30000");
  }

  init(): void {
    this.db.exec(SCHEMA);
  }

  close(): void {
    this.db.close();
  }

  insertTurn(input: InsertTurnInput): string {
    const turnId = newId("turn");
    const text = input.messages
      .map((m) => `${m.role ?? ""}: ${m.content ?? ""}`.trim())
      .join("\n")
      .trim();
    this.db
      .prepare(
        `INSERT INTO turns (id, session_id, user_id, timestamp, messages, metadata, text, created_at)
         VALUES (?,?,?,?,?,?,?,?)`,
      )
      .run(
        turnId,
        input.sessionId,
        input.userId,
        input.timestamp,
        JSON.stringify(input.messages),
        JSON.stringify(input.metadata),
        text,
        nowIso(),
      );
    return turnId;
  }

  /**
   * Persist an extracted memory with fact-evolution semantics.
   * - mutable slot: a different value supersedes the active one (active=0,
   *   supersedes -> old id); same value just bumps confidence/updated_at.
   * - additive slot: dedupe on (key, value); else insert.
   */
  addMemory(
    em: ExtractedMemory,
    ctx: { userId: string | null; sessionId: string; sourceTurn: string },
  ): string {
    const now = nowIso();
    const existing = this.db
      .prepare(
        "SELECT * FROM memories WHERE user_id IS ? AND key=? AND active=1 ORDER BY updated_at DESC",
      )
      .all(ctx.userId, em.key) as MemoryRow[];

    if (em.mutable) {
      const match = existing[0];
      if (match) {
        if (norm(match.value) === norm(em.value)) {
          this.db
            .prepare("UPDATE memories SET updated_at=?, confidence=MAX(confidence, ?) WHERE id=?")
            .run(now, em.confidence, match.id);
          return match.id;
        }
        this.db.prepare("UPDATE memories SET active=0, updated_at=? WHERE id=?").run(now, match.id);
        return this.insertMemory(em, ctx, now, match.id);
      }
    } else {
      for (const row of existing) {
        if (norm(row.value) === norm(em.value)) {
          this.db
            .prepare("UPDATE memories SET updated_at=?, confidence=MAX(confidence, ?) WHERE id=?")
            .run(now, em.confidence, row.id);
          return row.id;
        }
      }
    }
    return this.insertMemory(em, ctx, now, null);
  }

  private insertMemory(
    em: ExtractedMemory,
    ctx: { userId: string | null; sessionId: string; sourceTurn: string },
    now: string,
    supersedes: string | null,
  ): string {
    const id = newId("mem");
    this.db
      .prepare(
        `INSERT INTO memories (id, user_id, session_id, type, key, value, confidence,
           source_session, source_turn, created_at, updated_at, supersedes, active)
         VALUES (?,?,?,?,?,?,?,?,?,?,?,?,1)`,
      )
      .run(
        id,
        ctx.userId,
        ctx.sessionId,
        em.type,
        em.key,
        em.value,
        em.confidence,
        ctx.sessionId,
        ctx.sourceTurn,
        now,
        now,
        supersedes,
      );
    return id;
  }

  listMemories(userId: string | null, activeOnly: boolean): MemoryRow[] {
    let q = "SELECT * FROM memories WHERE user_id IS ?";
    if (activeOnly) q += " AND active=1";
    q += " ORDER BY created_at ASC";
    return this.db.prepare(q).all(userId) as MemoryRow[];
  }

  supersededValues(userId: string | null, key: string): string[] {
    const rows = this.db
      .prepare(
        "SELECT value FROM memories WHERE user_id IS ? AND key=? AND active=0 ORDER BY updated_at DESC",
      )
      .all(userId, key) as Array<{ value: string }>;
    return rows.map((r) => r.value);
  }

  recentTurns(args: {
    sessionId: string | null;
    userId: string | null;
    limit?: number;
  }): TurnRecord[] {
    const limit = args.limit ?? 50;
    let rows: Array<Record<string, unknown>>;
    if (args.sessionId !== null) {
      rows = this.db
        .prepare("SELECT * FROM turns WHERE session_id = ? ORDER BY created_at DESC LIMIT ?")
        .all(args.sessionId, limit) as Array<Record<string, unknown>>;
    } else if (args.userId !== null) {
      rows = this.db
        .prepare("SELECT * FROM turns WHERE user_id IS ? ORDER BY created_at DESC LIMIT ?")
        .all(args.userId, limit) as Array<Record<string, unknown>>;
    } else {
      return [];
    }
    return rows.map((r) => ({
      id: r.id as string,
      session_id: r.session_id as string,
      user_id: (r.user_id as string) ?? null,
      timestamp: (r.timestamp as string) ?? null,
      text: (r.text as string) ?? "",
      metadata: JSON.parse((r.metadata as string) || "{}"),
    }));
  }

  search(
    query: string,
    args: { userId: string | null; sessionId: string | null; limit?: number },
  ): SearchHit[] {
    const limit = args.limit ?? 10;
    const qset = tokenSet(query);
    const hits: SearchHit[] = [];

    const memRows = (
      args.userId !== null
        ? this.db.prepare("SELECT * FROM memories WHERE active=1 AND user_id IS ?").all(args.userId)
        : this.db.prepare("SELECT * FROM memories WHERE active=1").all()
    ) as MemoryRow[];
    for (const r of memRows) {
      const content = `${r.key}: ${r.value}`;
      const score = overlapScore(qset, content);
      if (score > 0) {
        hits.push({
          content,
          score,
          session_id: r.session_id,
          timestamp: r.updated_at,
          metadata: { type: r.type, memory_id: r.id },
        });
      }
    }

    let turnRows: Array<Record<string, unknown>> = [];
    if (args.sessionId !== null) {
      turnRows = this.db
        .prepare("SELECT * FROM turns WHERE session_id = ?")
        .all(args.sessionId) as Array<Record<string, unknown>>;
    } else if (args.userId !== null) {
      turnRows = this.db
        .prepare("SELECT * FROM turns WHERE user_id IS ?")
        .all(args.userId) as Array<Record<string, unknown>>;
    }
    for (const r of turnRows) {
      const text = (r.text as string) ?? "";
      const score = overlapScore(qset, text);
      if (score > 0) {
        hits.push({
          content: text,
          score,
          session_id: (r.session_id as string) ?? null,
          timestamp: (r.timestamp as string) ?? null,
          metadata: JSON.parse((r.metadata as string) || "{}"),
        });
      }
    }

    hits.sort((a, b) => b.score - a.score);
    return hits.slice(0, Math.max(0, limit));
  }

  deleteSession(sessionId: string): void {
    this.db.prepare("DELETE FROM turns WHERE session_id=?").run(sessionId);
    this.db.prepare("DELETE FROM memories WHERE session_id=?").run(sessionId);
  }

  deleteUser(userId: string): void {
    this.db.prepare("DELETE FROM turns WHERE user_id IS ?").run(userId);
    this.db.prepare("DELETE FROM memories WHERE user_id IS ?").run(userId);
  }
}
